### PR TITLE
Feature/top menu refactoring

### PR DESCRIPTION
**Pull request by GarstgerUnhold**
This does 2 things:
- refactors top menu javascript to get rid of its crazy ball-of-mud structure
- enables the project drop down to open links into new tabs as requested per https://www.openproject.org/issues/1695

### CODE REVIEW

**Comment by tessi** [comment]
the cukes seem to fail.

**Comment by GarstgerUnhold** [comment]
I threw out too much and broke the project menu. But things are better now, hooray!

File: app/assets/javascripts/top_menu.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    closeAllItems: function () {
+      var self = this;
+      this.openDropdowns().each(function (ix, item) {
+        self.close($(item));
+      });
+    },
+
+    closeOnBodyClick: function () {
+      var self = this;
+      $('html').click(function() {
```
Comment (nkenneweg) [inline app/assets/javascripts/top_menu.js:68]: I think you are adding this listener more than once but it is only needed once right?

File: app/assets/javascripts/top_menu.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      this.slideUp(dropdown);
+      dropdown.trigger("closed", dropdown);
+    },
+
+    closeOtherItems: function (dropdown) {
+      var self = this;
+      this.openDropdowns().filter(function (ix, it) {
+        return $(it) != $(dropdown);
+      }).each(function (ix, it) {
+        self.close($(it));
```
Comment (nkenneweg) [inline app/assets/javascripts/top_menu.js]: I guess you are using this filter and each stuff to make sure that the current dropdown really is one of our dropdowns.
Wouldn't it be easier to just do self.close($(dropdown))?

File: app/assets/javascripts/top_menu.js [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    },
+
+    // If there is ANY input, it will have precedence over links,
+    // i.e. links will only get focussed, if there is NO input whatsoever
+    focusFirstInputOrLink: function (dropdown) {
+      var toFocus = dropdown.find("ul :input:visible:first");
+      if (toFocus.length == 0) {
+        toFocus = dropdown.find("ul a:visible:first");
+      }
+      toFocus.blur();
```
Comment (nkenneweg) [inline app/assets/javascripts/top_menu.js:194]: I think this needs some explanation why you do not call focus() directly.

**Comment by nkenneweg** [comment]
I like the PR in general (+1 for refactoring spaghetti code)
Still have some smaller questions!

File: app/assets/javascripts/top_menu.js (hunk 2 above)
Comment (GarstgerUnhold) [inline app/assets/javascripts/top_menu.js]: Yup, it'd be easier. Gonna change that when I find the time.

File: app/assets/javascripts/top_menu.js (hunk 1 above)
Comment (GarstgerUnhold) [inline app/assets/javascripts/top_menu.js:68]: Yeah, could probably go into the setup thingy.

File: app/assets/javascripts/top_menu.js (hunk 3 above)
Comment (GarstgerUnhold) [inline app/assets/javascripts/top_menu.js:194]: As with any strange code, i.e. calls to timeout or multiple blurs, I'm gonna blame this on IE. I think there even was a comment above that in the Prototype version, seems like I threw it away out of sheer recklessness. Gonna put it there, again.

Comment (GarstgerUnhold) [inline app/assets/javascripts/top_menu.js:194]: Actually, this was the comment:
`// actually a simple focus should be enough.`
`// The rest is only there to work around a rendering bug in webkit (as of Oct 2011)`

I tried breaking things in Chrome and Safari. Chrome was fine but Safari had issues on the signup dropdown. I'm going to shorten the timeout though, as I found things too be much smoother without that 100ms and it should accomplish the exact same thing.

**Comment by GarstgerUnhold** [comment]
As you can see by 9d47004 the project/admin menu uses some of the same capabilities. At some point, those should go into another prototype, which both the top_menu and the project/admin menu can extend. This means, another huge chunk could easily be removed from application.js.

**Comment by hschink** [comment]
It is not possible to select projects from the top menu. :alien:

**Pull request merged**